File: CodingDojo/games/expansion/src/main/csharp/Command.cs
/*-
 * #%L
 * expansion - it's a dojo-like platform from developers to developers.
 * %%
 * Copyright (C) 2016 - 2020 Codenjoy
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */
﻿using System;
using System.Collections.Generic;
using System.Text;

namespace ExpansionBot
{
    public class Command
    {
        public struct Increase
        {
            public int Count;
            public Point Region;
        }

        public struct Movement
        {
            public int Count;
            public Direction Direction;
            public Point Region;
        }

        private readonly List<Increase> increases = new List<Increase>();
        private readonly List<Movement> movements = new List<Movement>();

        public void AddIncrease(int count, Point region)
        {
            Increase inc;
            inc.Count = count;
            inc.Region = region;
            increases.Add(inc);
        }

        public void AddMovement(int count, Point region, Direction direction)
        {
            Movement mov;
            mov.Count = count;
            mov.Region = region;
            mov.Direction = direction;
            movements.Add(mov);
        }

        public string GenerateCommand()
        {
            StringBuilder sb = new StringBuilder();
            sb.Append("{");
            sb.Append("\"movements\":[");
            for (int index = 0; index < movements.Count; index++)
            {
                if (index > 0) sb.Append(",");
                Movement movement = movements[index];
                sb.Append("{");
                sb.Append("\"count\":");
                sb.Append(movement.Count);
                sb.Append(",\"region\":{\"x\":");
                sb.Append(movement.Region.x);
                sb.Append(",\"y\":");
                sb.Append(movement.Region.y);
                sb.Append("},\"direction\":");
                sb.AppendFormat("\"{0}\"", DirectionToString(movement.Direction));
                sb.Append("}");
            }
            sb.Append("],");
            sb.Append("\"increase\":[");
            for (int index = 0; index < increases.Count; index++)
            {
                if (index > 0) sb.Append(",");
                Increase increase = increases[index];
                sb.Append("{");
                sb.Append("\"count\":");
                sb.Append(increase.Count);
                sb.Append(",\"region\":{\"x\":");
                sb.Append(increase.Region.x);
                sb.Append(",\"y\":");
                sb.Append(increase.Region.y);
                sb.Append("}}");
            }
            sb.Append("]");
            sb.Append("}");
            return sb.ToString();
        }

        private string DirectionToString(Direction direction)
        {
            switch (direction)
            {
                case Direction.Left: return "LEFT";
                case Direction.Down: return "DOWN";
                case Direction.Up: return "UP";
                case Direction.Right: return "RIGHT";
                case Direction.RightDown: return "RIGHT_DOWN";
                case Direction.RightUp: return "RIGHT_UP";
                case Direction.LeftDown: return "LEFT_DOWN";
                case Direction.LeftUp: return "LEFT_UP";
                default:
                    return string.Empty;
            }
        }
    }
}

File: CodingDojo/games/icancode/src/main/python/command.py
#!/usr/bin/env python3

###
# #%L
# Codenjoy - it's a dojo-like platform from developers to developers.
# %%
# Copyright (C) 2018 - 2021 Codenjoy
# %%
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public
# License along with this program.  If not, see
# <http://www.gnu.org/licenses/gpl-3.0.html>.
# #L%
###


_COMMANDS = dict(
    DIE="ACT(0)",

    LEFT="LEFT",
    RIGHT="RIGHT",
    UP="UP",
    DOWN="DOWN",

    JUMP="ACT(1)",
    JUMP_LEFT="ACT(1),LEFT",
    JUMP_RIGHT="ACT(1),RIGHT",
    JUMP_UP="ACT(1),UP",
    JUMP_DOWN="ACT(1),DOWN",

    PULL_LEFT="ACT(2),LEFT",
    PULL_RIGHT="ACT(2),RIGHT",
    PULL_UP="ACT(2),UP",
    PULL_DOWN="ACT(2),DOWN",

    FIRE_LEFT="ACT(3),LEFT",
    FIRE_RIGHT="ACT(3),RIGHT",
    FIRE_UP="ACT(3),UP",
    FIRE_DOWN="ACT(3),DOWN",

    NULL=""
)


class Command:

    def __init__(self, command):
        for key, value in _COMMANDS.items():
            if command == key or command == value:
                self._name = key
                self._command = value
                break
        else:
            raise ValueError("No Such Command: {}".format(command))

    def get_name(self):
        return self._name

    def get_command(self):
        return self._command

    def inverted(self):
        _inv_dir = None
        if self._name == 'LEFT':
            _inv_dir = Command
            "RIGHT"
        elif self._name == 'RIGHT':
            _inv_dir = Command
            "LEFT"
        elif self._name == 'UP':
            _inv_dir = Command
            "DOWN"
        elif self._name == 'DOWN':
            _inv_dir = Command
            "UP"
        else:
            _inv_dir = Command
            "NULL"
        return _inv_dir

    def to_string(self):
        return self._command

    def __eq__(self, other):
        return self._name == other._name and self._command == other._command


if __name__ == '__main__':
    raise RuntimeError("This module is not designed to be ran from CLI")
